api/v1alpha1/podnetworkchaos_types.go

```diff
@@ -55,18 +55,50 @@ type PodNetworkChaosSpec struct {
 	TrafficControls []RawTrafficControl `json:"tcs,omitempty"`
 }
 
+// IPSetType represents the type of IP set
+type IPSetType string
+
+const (
+	SetIPSet     IPSetType = "list:set"
+	NetPortIPSet IPSetType = "hash:net,port"
+	NetIPSet     IPSetType = "hash:net"
+)
+
 // RawIPSet represents an ipset on specific pod
 type RawIPSet struct {
 	// The name of ipset
 	Name string `json:"name"`
 
-	// The contents of ipset
+	IPSetType IPSetType `json:"ipsetType"`
+
+	// The contents of ipset.
+	// Only available when IPSetType is NetIPSet.
+	// +optional
 	Cidrs []string `json:"cidrs"`
 
+	// The contents of ipset.
+	// Only available when IPSetType is NetPortIPSet.
+	// +optional
+	CidrAndPorts []CidrAndPort `json:"cidrAndPorts"`
+
+	// The contents of ipset.
+	// Only available when IPSetType is SetIPSet.
+	// +optional
+	SetNames []string `json:"setNames"`
+
 	// The name and namespace of the source network chaos
 	RawRuleSource `json:",inline"`
 }
 
+// CidrAndPort represents CIDR and port pair
+type CidrAndPort struct {
+	Cidr string `json:"cidr"`
+
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	Port uint16 `json:"port"`
+}
+
```

Comment:
We should also append fields to introduce protocol, like `tcp`, `udp` or `icmp`.  When specify ipset rule without protocol, it would only specify with `tcp`.

Reply:
example usage  like `192.168.1.1,udp:53`

Reply:
This is something I'd rather avoid right now for a couple of reasons:

- This PR is already huge. Filtering by protocols needs more changes to IP sets and/or iptables rules.
- Adding a support for filtering by protocols is outside the scope of this PR.
- Defaulting to TCP would break backward compatibility. As of now Chaos Mesh is protocol agnostic.

Reply:
Got that!